gateway/src/dstack/gateway/core/store.py

```
@@ -0,0 +1,316 @@
+                    service.id,
+                    service.domain,
+                    service.auth,
+                )
+                stack.push_async_callback(
+                    supress_exc_async(self.nginx.unregister_domain, service.domain)
+                )
+
+                # Notify subscribers
+                for subscriber in self._subscribers:
+                    await subscriber.on_register(project, service)
+                    stack.push_async_callback(
+                        supress_exc_async(subscriber.on_unregister, project, service.id)
+                    )
+
+                # All fine, remove rollbacks
+                stack.pop_all()
+
+            self.services[service.id] = service
+            self.projects[project].add(service.id)
+
+        logger.info("%s: service %s (%s) is registered now", project, service.id, service.domain)
+
+    async def unregister_service(self, project: str, service_id: str):
+        async with self._lock:
+            if service_id not in self.projects[project]:
+                raise GatewayError(
+                    f"Service ID {service_id!r} is not registered in project {project!r}"
+                )
+            service = self.services[service_id]
+
+            logger.debug("%s: unregistering service %s (%s)", project, service_id, service.domain)
+
+            await asyncio.gather(
+                # Terminate all SSH tunnels
+                *(
+                    run_async(replica.ssh_tunnel.stop)
+                    for replica in service.replicas
+                    if replica.ssh_tunnel is not None
+                ),
+                # Unregister from nginx
+                self.nginx.unregister_domain(service.domain),
+                # Notify subscribers
+                *(
+                    subscriber.on_unregister(project, service.id)
+                    for subscriber in self._subscribers
+                ),
+                return_exceptions=True,
+            )
```

Since return_exceptions=True, all exceptions raised inside coroutines will not be reraised and, thus, silently ignored. I think we should at least log unexpected exceptions with error level.